dmoj/packet.py

```diff
@@ -162,15 +166,51 @@ def _read_single(self) -> dict:
             return json.loads(utf8text(packet))
 
     def run(self):
-        self._read_async()
+        threading.Thread(target=self._periodically_flush_testcase_queue).start()
+        self._read_forever()
 
     def disconnect(self):
         self.close()
         self.judge.terminate_grading()
         sys.exit(0)
 
-    def run_async(self):
-        threading.Thread(target=self._read_async).start()
+    def _flush_testcase_queue(self):
+        with self._testcase_queue_lock:
+            if not self._testcase_queue:
+                return
+
+            # Copy and release the lock quickly so that submission can continue grading.
+            testcase_queue = self._testcase_queue[:]
+            self._testcase_queue.clear()
+
+        self._send_packet({'name': 'test-case-status',
```

Comment:
Race condition: this has to be in the lock. Imagine the background thread flushing, and gets preempted here. Then, the main grading thread ends a batch, calling this function, finding nothing in the queue, sends the batch end, then this thread resumes.